### ISSUE TITLE
Invalid query generated using OracleDriver for filter with Option column

### ISSUE THREAD

**Issue by dycu** [slick/slick#1684]
While using Slick 3.1.1 with OracleDriver for the following query definition:
```scala
table1 joinRight table2 on (_.key1 === _.key2))
.filter {
case (_, table2ValueOption) => table2ValueOption.parameter.map(_ === searchingValue).getOrElse(false)
case _ => false.bind
}
.result
```

I've received the following sql query:
```sql
select ... where {fn ifnull((case when (s11."parameter" is null) then null else (case when (s11."parameter" = '{searchingValue}') then 1 else 0 end) end),0)}
```

and in result the sql error:
> [2017-01-23 09:31:31] [42000][920] ORA-00920: invalid relational operator

I've tried the same with both freeslick and the slick-extentions but there were no difference. Across the whole application we use DatabaseConfig[JdbcProfile] abstraction, having the exact driver defined in the configuration by using two parameters:
```
driver = "freeslick.OracleProfile$"
db {
  ...
  driver = "oracle.jdbc.OracleDriver"
}
```

I'm still not sure if that's some general problem with Slick support for Oracle database or I'm using it wrong.

**Comment by Alnaimi-** [comment]
Just ran into the same problem using Slick 3.2 with Oracle's odbc8 drivers. Did you manage to solve this issue?

Doing a bit of digging we found that running Connection.nativeSql on the generated SQL statement turns the `{fn ifnull ...}` into `NVL(case else)` Integer expression rather than a Boolean expression:

`LEFT OUTER JOIN ... ON NVL((CASE  WHEN (x96.x101 IS NOT NULL) THEN (CASE WHEN (x96.x102 = x104.x105) THEN 1 ELSE  END) ELSE NULL  END),0))`

the fix for above statement is to turn this into a Boolean expression by adding `LHS = 1` for example:

`LEFT OUTER JOIN ... ON NVL((CASE  WHEN (x96.x101 IS NOT NULL) THEN (CASE WHEN (x96.x102 = x104.x105) THEN 1 ELSE  END) ELSE NULL  END),0) = 1)`

**Comment by cvogt** [comment]
maybe something @trevorsibanda can look at for 3.3 (which we'll try to release much much more timely than 3.2)

**Comment by Alnaimi-** [comment]
Done some further testing; 

The problem was occurring when you perform a join operation followed by a `.getOrElse(_)` clause.

Seems like the scalar optimizer produces incorrect SQL statement for the Oracle drivers but doesn't affect the other databases.

Note: the getOrElse was used in the code as there were sequential left join operations.

**Comment by y0ke** [comment]
Any updates on this issue?

**Comment by kubukoz** [comment]
Hi, we just hit this again. The proposed fix in #2059 solves the issue. Can it be revisited in the near future?
